### PR TITLE
Fix picker deselecting filtered elements

### LINKED ISSUE #1816: Galerie-Element "vergißt" ausgewählte Bilder
Kann in der Demo nachgestellt werden:

1. Galerie-Element anlegen und Bilder aus einem Unterordner wählen. Dabei den Unterordner im Pfad fest einstellen.
2. Dateimanager aufrufen und einen anderen Pfad dort einstellen.
3. Zurück zum Galerie-Element wechseln und weitere Bilder aus dem anderen Pfad hinzufügen. Nach dem Klick auf Anwenden sind die Bilder aus dem Pfad in Punkt 1 verschwunden.

### PR DESCRIPTION
**Pull request by fritzmg**
| Q                | A
| -----------------| ---
| Fixed issues     | Fixes #1816
| Docs PR or issue | -

This PR fixes the picker deselecting elements that are not visible any more due to filtering. It injects the previously selected elements for the `Backend.openModalSelector` JavaScript, which can then add or remove elements.

### CODE REVIEW

File: core-bundle/src/Resources/public/core.js [hunk 1]
```diff
@@ -920,16 +920,31 @@ var Backend =
...
 			var frm = window.frames['simple-modal-iframe'],
-				val = [], ul, inp, field, act, it, i;
+				val = [], ul, inp, field, act, it, i, sIndex;
 			if (frm === undefined) {
 				alert('Could not find the SimpleModal frame');
 				return;
 			}
 			ul = frm.document.getElementById(opt.id);
+			// Load the previous values (#1816)
+			if (pickerValue = ul.get('data-picker-value')) {
```
Comment (ausi) [inline core-bundle/src/Resources/public/core.js:930]: `pickerValue` needs to be decleared otherwise this would be a global variable.

**Review by ausi — CHANGES_REQUESTED** [review]

File: core-bundle/src/Resources/public/core.js (hunk 1 above)
Comment (fritzmg) [reply core-bundle/src/Resources/public/core.js:930]: done 👍

File: core-bundle/src/Resources/public/core.js [hunk 2]
```diff
@@ -920,16 +920,31 @@ var Backend =
...
 			for (i=0; i<inp.length; i++) {
-				if (inp[i].checked && !inp[i].id.match(/^(check_all_|reset_)/)) {
-					val.push(inp[i].get('value'));
+				if (inp[i].id.match(/^(check_all_|reset_)/)) {
+					continue;
+				}
+				// Add currently selected value, otherwise remove (#1816)
+				sIndex = val.indexOf(inp[i].get('value'));
+				if (inp[i].checked) {
+					if (sIndex < 0) {
```
Comment (ausi) [inline core-bundle/src/Resources/public/core.js]: Everywhere else in the the JavaScript we compare `indexOf` to `-1` maybe change for consistency?
```suggestion
					if (sIndex === -1) {
```

**Review by ausi — DISMISSED** [review]

File: core-bundle/src/Resources/public/core.js [hunk 3]
```diff
@@ -920,16 +920,31 @@ var Backend =
...
+						val.push(inp[i].get('value'));
+					}
+				} else {
+					if (sIndex >= 0) {
+						val.splice(sIndex, 1);
+					}
```
Comment (leofeyer) [inline core-bundle/src/Resources/public/core.js]:
```suggestion
				} else if (sIndex >= 0) {
					val.splice(sIndex, 1);
```

File: core-bundle/src/Resources/public/core.js (hunk 2 above)
Comment (leofeyer) [reply core-bundle/src/Resources/public/core.js:941]: Mostly we only check `== -1` and not `=== -1`. 🙈

**Review by leofeyer — CHANGES_REQUESTED** [review]

File: core-bundle/src/Resources/public/core.js (hunk 3 above)
Comment (fritzmg) [reply core-bundle/src/Resources/public/core.js]: imho the code is more readable without `else if`. Before:

```js
if (inp[i].checked) {
    if (sIndex < 0) {
        val.push(inp[i].get('value'));
    }
} else {
    if (sIndex >= 0) {
        val.splice(sIndex, 1);
    }
}
```

after:

```js
if (inp[i].checked) {
    if (sIndex < 0) {
        val.push(inp[i].get('value'));
    }
} else if (sIndex >= 0) {
    val.splice(sIndex, 1);
}
````

Even though both the `val.push` and the `val.splice` require two conditions to be executed (`checked && sIndex === -1` and `!checked && sIndex !== -1`), they are now on different indentation levels.

Comment (m-vo) [reply core-bundle/src/Resources/public/core.js]: > imho the code is more readable without else if

I tend to disagree. Imo both `if`s should also be consolidated.

Comment (fritzmg) [reply core-bundle/src/Resources/public/core.js]: Like this you mean? fe1beb1cde01a8907f51da63d93777c56ebe2b52

**Review by ausi — DISMISSED** [review]

**Review by aschempp — COMMENTED** [review]
are you correctly handling radio buttons (where there can only be one selection)?

**Review by leofeyer — APPROVED** [review]

**Comment by leofeyer** [comment]
@fritzmg What about @aschempp's [comment](https://github.com/contao/contao/pull/2296#pullrequestreview-490358335) above?

**Comment by fritzmg** [comment]
I'll check :)

**Comment by fritzmg** [comment]
> are you correctly handling radio buttons (where there can only be one selection)?

No special handling is required there as far as I tested. If the `fieldType` is `radio` and `multiple` is `false`, it works the same way. The selection does not get lost when filtering into a tree that is different from the current selection, not selecting anything there and then applying the (still previous) selection.

**Comment by fritzmg** [comment]
@leofeyer should I create a separate PR for Contao 4.9?

**Comment by leofeyer** [comment]
I don't think that is necessary.

**Comment by leofeyer** [comment]
Thank you @fritzmg.

**Pull request merged**